### PR TITLE
expression: fix cast invalid string to datetime

### LINKED ISSUE #24997: executor: select with `inl_hash_join` return inconsistency with MySQL
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

table
```
CREATE TABLE `tbl_2` (
  `col_10` time NOT NULL DEFAULT '05:44:48',
  `col_11` time DEFAULT '05:53:29',
  `col_12` date DEFAULT '2031-07-19',
  `col_13` date DEFAULT NULL,
  `col_14` date NOT NULL,
  PRIMARY KEY (`col_10`),
  UNIQUE KEY `idx_18` (`col_10`,`col_11`),
  UNIQUE KEY `idx_20` (`col_10`,`col_11`,`col_12`),
  UNIQUE KEY `idx_21` (`col_10`,`col_11`),
  UNIQUE KEY `idx_22` (`col_10`,`col_11`,`col_12`,`col_13`,`col_14`),
  KEY `idx_17` (`col_10`,`col_11`,`col_12`,`col_13`),
  KEY `idx_19` (`col_10`,`col_11`),
  KEY `idx_23` (`col_10`,`col_11`,`col_12`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci

CREATE TABLE `tbl_14` (
  `col_80` datetime DEFAULT NULL,
  `col_81` bit(1) DEFAULT b'0',
  `col_82` tinyint(1) DEFAULT NULL,
  `col_83` datetime NOT NULL,
  `col_84` tinyblob,
  `col_85` enum('Alice','Bob','Charlie','David') COLLATE utf8_unicode_ci DEFAULT NULL,
  `col_86` varchar(86) CHARACTER SET utf8mb4 COLLATE utf8mb4_unicode_ci DEFAULT '',
  `col_87` varbinary(305) DEFAULT NULL,
  `col_88` int DEFAULT '155374992',
  UNIQUE KEY `idx_71` (`col_80`,`col_81`,`col_82`,`col_83`,`col_88`),
  UNIQUE KEY `idx_72` (`col_80`,`col_81`,`col_82`,`col_83`,`col_88`)
) ENGINE=InnoDB DEFAULT CHARSET=utf8 COLLATE=utf8_unicode_ci

```

data
[tbl2.txt](https://github.com/pingcap/tidb/files/6569416/tbl2.txt)
[tbl14.txt](https://github.com/pingcap/tidb/files/6569417/tbl14.txt)

### 2. What did you expect to see? (Required)
result shoul d be the same since meta and data is the same

### 3. What did you see instead (Required)
ingest compare fail

### 4. What is your TiDB version? (Required)

master

### PR DESCRIPTION
**Pull request by wzru**
### What problem does this PR solve?

Issue Number: close #24997  

Problem Summary:

Simplified case:
```sql
drop table if exists t1, t2;
create table t1(c1 varbinary(100));
insert into t1 values('e');
create table t2(c1 date);
insert into t2 values('2019-11-10'); 
select * from t1 inner join t2 on t1.c1 <= t2.c1;
```
MySQL:
```sql
mysql> select * from t1 inner join t2 on t1.c1 <= t2.c1;
   
+------------+------------+
| c1         | c1         |
+------------+------------+
| 0x65       | 2019-11-10 |
+------------+------------+
1 row in set, 1 warning (0.00 sec)
```
TiDB:
```sql
mysql> select * from t1 inner join t2 on t1.c1 <= t2.c1;

Empty set, 1 warning (0.00 sec)
```
### What is changed and how it works?

What's Changed:

`func (b *builtinCastStringAsTimeSig) vecEvalTime`
`func (b *builtinCastStringAsTimeSig) evalTime`

How it Works:

### Check List 

Tests 

- [ ] Unit test
- [ ] Integration test

Side effects

- [ ] Performance regression: Consumes more CPU

### Release note 

```release-note
None
```

### CODE REVIEW

**Review by wshwsh12 — COMMENTED** [review]
Please add some test cases for your modify.
For example.
```
MySQL [test]> select cast('e' as date);
+-------------------+
| cast('e' as date) |
+-------------------+
| NULL              |
+-------------------+
```

**Comment by XuHuaiyu** [comment]
@wzru 
Any update for this PR?

File: expression/builtin_cast_vec.go [hunk 1]
```diff
@@ -1696,8 +1696,6 @@ func (b *builtinCastStringAsTimeSig) vecEvalTime(input *chunk.Chunk, result *chu
 			if err = handleInvalidTimeError(b.ctx, err); err != nil {
 				return err
 			}
-			result.SetNull(i, true)
```
Comment (wshwsh12) [inline expression/builtin_cast_vec.go:1699]: Should we use b.ctx.GetSessionVars().StrictSQLMode?

**Comment by hawkingrei** [comment]
@wzru Do you have time to finish this PR? If there is no reply within a week, we will close the PR and take over the job.

**Comment by wzru** [comment]
@hawkingrei Sorry for that. I don't have time yet. I will close this PR.

**Comment by hawkingrei** [comment]
> @hawkingrei Sorry for that. I don't have time yet. I will close this PR.

Thank you for your contribution.

**Pull request closed without merging**